File: BeaconActiveSDKData-ObjC/SensoroSDKActiveData-ObjC/SENLocationManager.m
//
//  SENLocationManager.m
//  BeaconActive-ObjC
//
//  Created by skyming on 15/4/10.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

#import "SENLocationManager.h"
#import <UIKit/UIKit.h>
#import <SensoroBeaconKit/SensoroBeaconKit.h>

@interface SENLocationManager ()<SBKBeaconManagerDelegate>

@end

@implementation SENLocationManager

+ (instancetype)sharedInstance
{
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init{
    
    if(self = [super init])
    {
        _locationManager = [SBKBeaconManager sharedInstance];

        if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 8.0)
        {
            //获取授权
            [_locationManager requestAlwaysAuthorization];
        }
        //设定代理
        _locationManager.delegate = self;
        //生成监测用的区域ID
        _monitorRegion = [SBKBeaconID beaconIDWithProximityUUID:[[NSUUID alloc]initWithUUIDString:@"46D06053-9FAD-483B-B704-E576735CE1A4"] major:0x23A2 minor:0x854D];
        
    }
    return self;
}


- (void)startMonitor:(BOOL)relaunch{
    
    if (!relaunch) {
        [_locationManager startRangingBeaconsWithID:_monitorRegion wakeUpApplication:YES];
        NSLog(@"Start monitor region!");
    }else{
        NSLog(@"During the relauch app, don't restart monitor region!");
        [_locationManager startRangingBeaconsWithID:_monitorRegion wakeUpApplication:NO];

    }
    _started = true;

}

- (void)stopMonitor{
    [_locationManager stopRangingBeaconsWithID:_monitorRegion];
    _started = false;
    NSLog(@"Stop monitor region!");
}

- (void)sendNotification:(NSString *)notify{
    UILocalNotification *notification = [[UILocalNotification alloc]init];
    notification.alertBody = notify;
    notification.soundName = UILocalNotificationDefaultSoundName;
    
    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
}


- (void)beaconManager:(SBKBeaconManager *)beaconManager didRangeNewBeacon:(SBKBeacon *)beacon{
   
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    NSString *alertbody =[NSString stringWithFormat:@"Enter: %@ %@ %@ %@ %@ %@ %@",beacon.beaconID.proximityUUID.UUIDString, beacon.beaconID.major,beacon.beaconID.minor,beacon.batteryLevel,beacon.temperature,beacon.light,beacon.accelerometerCount];
    [self sendNotification:alertbody];
    
    NSLog(@"Enter: %@",now);
}

- (void)beaconManager:(SBKBeaconManager *)beaconManager beaconDidGone:(SBKBeacon *)beacon{
  
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    NSString *alertbody =[NSString stringWithFormat:@"Exit: %@ %@ %@ %@ %@ %@ %@",beacon.beaconID.proximityUUID.UUIDString, beacon.beaconID.major,beacon.beaconID.minor,beacon.batteryLevel,beacon.temperature,beacon.light,beacon.accelerometerCount];
    [self sendNotification:alertbody];

    NSLog(@"Exit: %@",now);
}

- (void)beaconManager:(SBKBeaconManager *)beaconManager didDetermineState:(SBKRegionState)state forRegion:(SBKBeaconID *)region{
  
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    switch (state) {
        case SBKRegionStateEnter:
        {
            NSString *alertbody =[NSString stringWithFormat:@"Enter region at %@",[formatter stringFromDate:now]];
            [self sendNotification:alertbody];
            NSLog(@"%@",alertbody);
        }
            break;
            
        case SBKRegionStateLeave:
        {
            NSString *alertbody =[NSString stringWithFormat:@"Exit region at %@",[formatter stringFromDate:now]];
            [self sendNotification:alertbody];
            NSLog(@"%@",alertbody);
        }
            break;
            
        case SBKRegionStateUnknown:
            NSLog(@"This region state was unknown!");
            break;
    }

    if ([UIApplication sharedApplication].applicationState != UIApplicationStateActive) {
        __block UIBackgroundTaskIdentifier bgTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
            // Clean up any unfinished task business by marking where you
            // stopped or ending the task outright.
            [self sendNotification:@"后台任务结束"];
            [[UIApplication sharedApplication] endBackgroundTask:bgTask];
            bgTask = UIBackgroundTaskInvalid;
        }];
    }

}

- (void)beaconManager:(SBKBeaconManager *)beaconManager scanDidFinishWithBeacons:(NSArray *)beacons{
    
    NSLog(@"应用已被激活 %d",(int)beacons.count);
    
    SBKBeacon *beacon = nil;
    if (beacons.count > 0) {
        beacon = beacons[0];
        NSLog(@"Beacon: %@ %@", beacon,beacon.beaconID.debugDescription);
        NSString *alertbody =[NSString stringWithFormat:@"Beacon %@ %@ %@ %@ %@ %@ %@",beacon.beaconID.proximityUUID.UUIDString, beacon.beaconID.major,beacon.beaconID.minor,beacon.batteryLevel,beacon.temperature,beacon.light,beacon.accelerometerCount];
//        [self sendNotification:alertbody];
    }
    
}


@end

File: BeaconActiveSDKData-Swift/SensoroSDKActive-iOS/SENLocationManager.swift
//
//  SENLocationManager.swift
//  SensoroSDKActive-iOS
//
//  Created by David Yang on 15/4/8.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

import UIKit
import CoreLocation
import SensoroBeaconKit

private let _sharedInstance = SENLocationManager();

class SENLocationManager: NSObject, SBKBeaconManagerDelegate {
    
    class var sharedInstance : SENLocationManager {
        return _sharedInstance;
    }

    var started = false;
    var monitorRegion : SBKBeaconID!;

    var alertbody: NSMutableString? = nil;
    
    private override init(){
        
        super.init();

        //设定代理
        SBKBeaconManager.sharedInstance().delegate = self;
        //获取授权
        SBKBeaconManager.sharedInstance().requestAlwaysAuthorization();
        //生成监测用的区域ID
        monitorRegion = SBKBeaconID(proximityUUID: NSUUID(UUIDString: "46D06053-9FAD-483B-B704-E576735CE1A3"));
    }
    
    func startMonitor( relaunch : Bool ){
        if relaunch == false {
            SBKBeaconManager.sharedInstance().startRangingBeaconsWithID(monitorRegion, wakeUpApplication: true);
            NSLog("Start monitor region!");
        }else{
            NSLog("During the relauch app, don't restart monitor region!");
            SBKBeaconManager.sharedInstance().startRangingBeaconsWithID(monitorRegion, wakeUpApplication: false);
        }
        started = true;
    }
    
    
    func stopMonitor(){
        SBKBeaconManager.sharedInstance().stopRangingBeaconsWithID(monitorRegion);
        started = false;
        NSLog("Stop monitor region!");
    }
    
    func sendNotification(notify : String){
        var notification = UILocalNotification()
        notification.alertBody = notify
        
        notification.soundName = UILocalNotificationDefaultSoundName
        UIApplication.sharedApplication().scheduleLocalNotification(notification)
    }
    
    func checkUnwrapping(sensorData:NSNumber?){
        
        if let info = sensorData{
            alertbody?.appendFormat("% @ ", sensorData!);
        }
    }
    
    // MARK: SBKBeaconManagerDelegate
    func beaconManager(beaconManager: SBKBeaconManager!, didRangeNewBeacon beacon: SBKBeacon!) {
        let now = NSDate();
        let formatter = NSDateFormatter();
        formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
        //NSLog("Enter a beacon at \(formatter.stringFromDate(now))");
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, beaconDidGone beacon: SBKBeacon!) {
        let now = NSDate();
        let formatter = NSDateFormatter();
        formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
        //NSLog("Leave a region at \(formatter.stringFromDate(now))");
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, scanDidFinishWithBeacons beacons: [AnyObject]!) {
        //NSLog("Periodically call for scanDidFinishWithBeacons");
        if(beacons.count > 0){
            var beacon:SBKBeacon! = beacons[0] as? SBKBeacon;
            alertbody = NSMutableString(format: "Beacon %@ %@ %@ ", beacon!.beaconID.proximityUUID.UUIDString,beacon!.beaconID.major,beacon!.beaconID.minor);
         
            self.checkUnwrapping(beacon!.batteryLevel);
            self.checkUnwrapping(beacon!.temperature);
            self.checkUnwrapping(beacon!.light);
            self.checkUnwrapping(beacon!.accelerometerCount);
            
            self.sendNotification("\(alertbody)");
        }
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, didDetermineState state : SBKRegionState, forRegion region : SBKBeaconID!) {
        
        switch state {
        case .Enter:
            let now = NSDate();
            let formatter = NSDateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification("Enter region at \(formatter.stringFromDate(now))");
            NSLog("Enter region at \(formatter.stringFromDate(now))");
        case .Leave:
            let now = NSDate();
            let formatter = NSDateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification("Exit  region at \(formatter.stringFromDate(now))");
            NSLog("Exit  region at \(formatter.stringFromDate(now))");
        case .Unknown:
            NSLog("This region state was unknown!");
        }
        
        var bgTask : UIBackgroundTaskIdentifier? = nil;
        bgTask = UIApplication.sharedApplication().beginBackgroundTaskWithExpirationHandler { () -> Void in
            self.sendNotification("后台任务结束");
            if let bgTaskInner = bgTask {
                UIApplication.sharedApplication().endBackgroundTask(bgTaskInner);
                bgTask = UIBackgroundTaskInvalid;
            }
        }

    }
    
    
}


